crates/aptos-api-tester/src/main.rs

```diff
@@ -0,0 +1,250 @@
+    };
+
+    // check account data
+    let expected_account = Account {
+        authentication_key: account.authentication_key(),
+        sequence_number: account.sequence_number(),
+    };
+    let actual_account = response.inner();
+
+    if expected_account.authentication_key != actual_account.authentication_key {
+        return TestResult::Fail("wrong authentication key");
+    }
+    if expected_account.sequence_number != actual_account.sequence_number {
+        return TestResult::Fail("wrong sequence number");
+    }
+
+    // check account balance
+    let expected_balance = U64(amount_funded);
+    let actual_balance = match client.get_account_balance(account.address()).await {
+        Ok(response) => response.inner().coin.value,
+        Err(e) => return TestResult::Error(e.into()),
+    };
+
+    if expected_balance != actual_balance {
+        return TestResult::Fail("wrong balance");
+    }
+
+    TestResult::Success
+}
+
+/// Tests coin transfer. Checks that:
+///   - receiver balance reflects transferred amount
+///   - receiver balance shows correct amount at the previous version
+async fn test_cointransfer(
+    client: &Client,
+    coin_client: &CoinClient<'_>,
+    account: &mut LocalAccount,
+    address: AccountAddress,
+    amount: u64,
+) -> TestResult {
+    // get starting balance
+    let starting_receiver_balance = match client.get_account_balance(address).await {
+        Ok(response) => u64::from(response.inner().coin.value),
+        Err(e) => return TestResult::Error(e.into()),
+    };
+
+    // transfer coins to static account
+    let txn_hash = match coin_client.transfer(account, address, amount, None).await {
+        Ok(txn) => txn,
+        Err(e) => return TestResult::Error(e),
+    };
+    let response = match client.wait_for_transaction(&txn_hash).await {
+        Ok(response) => response,
+        Err(e) => return TestResult::Error(e.into()),
+    };
+
+    // check receiver balance
+    let expected_receiver_balance = U64(starting_receiver_balance + amount);
+    let actual_receiver_balance = match client.get_account_balance(address).await {
+        Ok(response) => response.inner().coin.value,
+        Err(e) => return TestResult::Error(e.into()),
+    };
+
+    if expected_receiver_balance != actual_receiver_balance {
+        return TestResult::Fail("wrong balance after coin transfer");
+    }
+
+    // check account balance with a lower version number
+    let version = match response.inner().version() {
+        Some(version) => version,
+        _ => return TestResult::Error(anyhow!("transaction did not return version")),
+    };
+
+    let expected_balance_at_version = U64(starting_receiver_balance);
+    let actual_balance_at_version = match client
+        .get_account_balance_at_version(address, version - 1)
+        .await
+    {
+        Ok(response) => response.inner().coin.value,
+        Err(e) => return TestResult::Error(e.into()),
+    };
```

Comment:
You can cut this down with a `.map_err(|err| TestResult::Error(err.into()))?.inner().coin.value` or similar if you want.

Reply:
I think this returns an Err(TestResult::Error)?

Reply:
Oh, yes you're right since it's an enum type.

The ? is only applicable to Result, so it requires reworking some other pieces then.  So, this is fine for now.